pkg/controllers/node/node_controller.go

```diff
@@ -159,139 +104,112 @@ func (c *NodeController) Run(threadiness int, reconcilerPeriod string, stopCh ch
 	log.Debug("Finished syncing with Kubernetes API (Nodes)")
 
 	// Start Calico cache.
-	c.k8sResourceCache.Run(reconcilerPeriod)
+	go c.acceptScheduleRequests(stopCh)
 
-	// Start a number of worker threads to read from the queue.
-	for i := 0; i < threadiness; i++ {
-		go c.runWorker()
-	}
 	log.Info("Node controller is now running")
 
+	// Kick off a start of day sync.
+	c.schedule <- nil
+
 	<-stopCh
 	log.Info("Stopping Node controller")
 }
 
-func (c *NodeController) runWorker() {
-	for c.processNextItem() {
+// acceptScheduleRequests monitors the schedule channel for kicks to wake up
+// and schedule syncs.
+func (c *NodeController) acceptScheduleRequests(stopCh <-chan struct{}) {
+	for {
+		// Wait until something wakes us up, or we are stopped
+		select {
+		case <-c.schedule:
+			c.doSchedule(stopCh)
+		case <-stopCh:
+			return
+		}
 	}
 }
 
-func (c *NodeController) processNextItem() bool {
-	// Wait until there is a new item in the work queue.
-	workqueue := c.k8sResourceCache.GetQueue()
-	key, quit := workqueue.Get()
-	if quit {
-		return false
+// doSchedule actually performs the scheduling of syncs. It is a separate method
+// so that we don't introduce locking into the acceptScheduleRequests method.
+func (c *NodeController) doSchedule(stopCh <-chan struct{}) {
```

Comment:
IIUC this code correctly, then we want to trigger a reconciliation upon a delete, but to consolidate delete events while a reconciliation is happening into a single trigger event.

Funnily enough I did something very similar today to an asynchronous write to file. The way I did this was:
-  Use a channel of length 1 as a trigger channel
-  Reconciliation would loop indefinitely waiting for an event on  the trigger channel and performing a reconciliation as soon as it gets one.
-  Delete events from the k8s informer simply use a non blocking send to push an item into the trigger channel.  If there is already an entry in the channel then the non-blocking send doesn't do anything.

This approach doesn't require any locking and feels a little more go-ey.

Reply:
Nice! That simplifies quite a bit